algebird-core/src/main/scala/com/twitter/algebird/QTree.scala

```diff
@@ -108,32 +152,45 @@ class QTreeSemigroup[A](k: Int)(implicit val underlyingMonoid: Monoid[A]) extend
   }
 }
 
-case class QTree[A](
+case class QTree[@specialized(Int, Long, Float, Double, Unit) A] private[algebird] (
   offset: Long, //the range this tree covers is offset*(2^level) ... (offset+1)*(2^level)
   level: Int,
   count: Long, //the total count for this node and all of its children
   sum: A, //the sum at just this node (*not* including its children)
-  lowerChild: Option[QTree[A]],
-  upperChild: Option[QTree[A]]) {
+  private[algebird]lowerChildNullable: QTree[A],
+  private[algebird]upperChildNullable: QTree[A]) {
+  import QTree._
 
-  require(offset >= 0, "QTree can not accept negative values")
+  val range: Double =
+    if (level < cachedRangeCacheSize && level > cachedRangeLowerBound)
+      rangeLut(level + cachedRangeCacheSize)
+    else
+      math.pow(2.0, level)
 
-  def range: Double = math.pow(2.0, level)
   def lowerBound: Double = range * offset
   def upperBound: Double = range * (offset + 1)
 
-  private def extendToLevel(n: Int)(implicit monoid: Monoid[A]): QTree[A] = {
+  def lowerChild: Option[QTree[A]] = Option(lowerChildNullable)
+  def upperChild: Option[QTree[A]] = Option(upperChildNullable)
+
+  @annotation.tailrec
+  private[algebird] final def extendToLevel(n: Int)(implicit monoid: Monoid[A]): QTree[A] = {
     if (n <= level)
       this
     else {
       val nextLevel = level + 1
       val nextOffset = offset / 2
 
+      var l: QTree[A] = null
```

Comment:
can we add a comment that this was due to benchmarking and not to change it unless you have run benchmarks to show that it is faster.